### PR TITLE
Convey - DO NOT MERGE

### PR DESCRIPTION
**Pull request by kumarad**

### CODE REVIEW

File: java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/StreamUtil.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import java.io.InputStream;
+import java.security.DigestInputStream;
+import java.security.MessageDigest;
+
+/**
+ * Utility to perform stream operations using chunks to save on JVM memory.
+ */
+public class StreamUtil {
+
+    private static final int CHUNK_SIZE = 64 * 1034;
```
Comment (clintmiller1) [inline java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/StreamUtil.java]: Where does 1034 come from? Should this be a power of 2?

File: java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java [hunk 2]
```diff
@@ -887,21 +880,17 @@ private String extractBinaries() {
...
                     try (FileOutputStream lockFOS = new FileOutputStream(lockFile);
                          FileLock lock = lockFOS.getChannel().lock()) {
+                        InputStream binaryInputStream = getBinaryInputStream(root, os, executableName);
                         if (extracted.exists()) {
-                            boolean contentEqual = false;
-                            if (extracted.length() == bin.length) {
-                                try (InputStream executableIS = new FileInputStream(extracted)) {
-                                    byte[] existingBin = IOUtils.toByteArray(executableIS);
-                                    contentEqual = Arrays.equals(bin, existingBin);
+                            try (InputStream executableIS = new FileInputStream(extracted)) {
```
Comment (clintmiller1) [inline java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java]: I think executableIS is being closed twice. The try() statement above closes it. But StreamUtil.compareStreamsChunked() also closes it. I'm thinking that compareStreamsChunked() should not be closing it.

File: java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java [hunk 3]
```diff
@@ -887,21 +880,17 @@ private String extractBinaries() {
                     final File lockFile = new File(pathToLock);
                     try (FileOutputStream lockFOS = new FileOutputStream(lockFile);
                          FileLock lock = lockFOS.getChannel().lock()) {
+                        InputStream binaryInputStream = getBinaryInputStream(root, os, executableName);
```
[the remark this answers is not in the document]
Comment (clintmiller1) [reply java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java]: So I guess there are 2 paths. If the output file already exists, we go through compareStreamsChunked() which closes the stream. On this path there's a small window where an exception could leave the stream open.

The other path is the IOUtils.copyLarge() path. I think when we go down that path, the stream isn't being closed at all.

File: java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java [hunk 4]
```diff
@@ -873,11 +869,8 @@ private String extractBinaries() {
                     
                     String extension = os.equals("windows") ? ".exe" : "";
                     String executableName = "kinesis_producer" + extension;
-                    byte[] bin = IOUtils.toByteArray(
-                            this.getClass().getClassLoader().getResourceAsStream(root + "/" + os + "/" + executableName));
-                    MessageDigest md = MessageDigest.getInstance("SHA1");
-                    String mdHex = DatatypeConverter.printHexBinary(md.digest(bin)).toLowerCase();
-                    
+                    String mdHex = StreamUtil.getMD5(getBinaryInputStream(root, os, executableName));
```
[the remark this answers is not in the document]
Comment (kumarad) [reply java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java]: its closed by getMD5 to try and keep things simple.

File: java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/StreamUtil.java (hunk 1 above)
Comment (kumarad) [reply java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/StreamUtil.java]: thats a typo. will update to 1024.

File: java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (kumarad) [reply java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java]: you are right. this was an existing whole. but you are right about us probably wanting to fix it.

**Pull request closed without merging**